File: Cardiovascular Disease Prediction using Machine Learning and Clinical Data/scripts/r/univariate_analysis.R
# Univariate Analysis
#
# This script performs univariate analysis on the Cardiovascular Disease Dataset.

# Load necessary libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(psych)
library(gridExtra)

# Set theme for plots
theme_set(theme_minimal() + theme(plot.title = element_text(size = 14, face = "bold")))

# Load the dataset
df <- read.csv("../../data/Cardiovascular_Disease_Dataset.csv", stringsAsFactors = FALSE)

# Univariate Analysis for Numerical Variables
cat(paste(rep("=", 80), collapse = ""), "\n")
cat("UNIVARIATE ANALYSIS - NUMERICAL VARIABLES\n")
cat(paste(rep("=", 80), collapse = ""), "\n")

numerical_cols <- c("age", "restingBP", "serumcholestrol", "maxheartrate", "oldpeak")

for (col in numerical_cols) {
  if (col %in% names(df)) {
    cat("\n", toupper(col), ":\n")
    cat("  Mean:", mean(df[[col]], na.rm = TRUE), "\n")
    cat("  Median:", median(df[[col]], na.rm = TRUE), "\n")
    cat("  Std Dev:", sd(df[[col]], na.rm = TRUE), "\n")
    cat("  Min:", min(df[[col]], na.rm = TRUE), "\n")
    cat("  Max:", max(df[[col]], na.rm = TRUE), "\n")
    cat("  Skewness:", psych::skew(df[[col]], na.rm = TRUE), "\n")
    cat("  Kurtosis:", psych::kurtosi(df[[col]], na.rm = TRUE), "\n")
    
    # Outlier detection
    Q1 <- quantile(df[[col]], 0.25, na.rm = TRUE)
    Q3 <- quantile(df[[col]], 0.75, na.rm = TRUE)
    IQR_val <- Q3 - Q1
    outliers <- df[df[[col]] < Q1 - 1.5 * IQR_val | df[[col]] > Q3 + 1.5 * IQR_val, ]
    cat("  Outliers:", nrow(outliers), "(", round(nrow(outliers)/nrow(df)*100, 2), "%)\n")
  }
}

# Univariate Analysis for Categorical Variables
cat("\n\n", paste(rep("=", 80), collapse = ""), "\n")
cat("UNIVARIATE ANALYSIS - CATEGORICAL VARIABLES\n")
cat(paste(rep("=", 80), collapse = ""), "\n")

categorical_cols <- c("gender", "chestpain", "fastingbloodsugar", "restingrelectro", 
                     "exerciseangia", "slope", "noofmajorvessels", "target")

for (col in categorical_cols) {
  if (col %in% names(df)) {
    cat("\n", toupper(col), ":\n")
    value_counts <- table(df[[col]])
    print(value_counts)
    cat("Percentage:\n")
    print(prop.table(value_counts) * 100)
  }
}

# Visualize distributions
plots_list <- list()
for (col in numerical_cols) {
  if (col %in% names(df)) {
    p <- ggplot(df, aes_string(x = col)) +
      geom_histogram(bins = 30, fill = "steelblue", alpha = 0.7, color = "black") +
      labs(title = paste("Histogram:", col),
           x = col,
           y = "Frequency") +
      theme_minimal()
    plots_list[[col]] <- p
  }
}

# Save plots
png("../../results/univariate_analysis.png", width = 1200, height = 800)
do.call(grid.arrange, c(plots_list, ncol = 2))
dev.off()

cat("\nUnivariate analysis complete!\n")

File: Cardiovascular Disease Prediction using Machine Learning and Clinical Data/scripts/python/univariate_analysis.py
"""
Univariate Analysis

This script performs univariate analysis on the Cardiovascular Disease Dataset.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Set style
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

def load_data():
    """Load the dataset"""
    df = pd.read_csv('../../data/Cardiovascular_Disease_Dataset.csv')
    return df

def univariate_numerical(df):
    """Univariate analysis for numerical variables"""
    numerical_cols = ['age', 'restingBP', 'serumcholestrol', 'maxheartrate', 'oldpeak']
    
    print("=" * 80)
    print("UNIVARIATE ANALYSIS - NUMERICAL VARIABLES")
    print("=" * 80)
    
    for col in numerical_cols:
        if col in df.columns:
            print(f"\n{col.upper()}:")
            print(f"  Mean: {df[col].mean():.2f}")
            print(f"  Median: {df[col].median():.2f}")
            print(f"  Std Dev: {df[col].std():.2f}")
            print(f"  Min: {df[col].min():.2f}")
            print(f"  Max: {df[col].max():.2f}")
            print(f"  Skewness: {df[col].skew():.4f}")
            print(f"  Kurtosis: {df[col].kurtosis():.4f}")
            
            # Outlier detection
            Q1 = df[col].quantile(0.25)
            Q3 = df[col].quantile(0.75)
            IQR = Q3 - Q1
            outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
            print(f"  Outliers: {len(outliers)} ({len(outliers)/len(df)*100:.2f}%)")

def univariate_categorical(df):
    """Univariate analysis for categorical variables"""
    categorical_cols = ['gender', 'chestpain', 'fastingbloodsugar', 'restingrelectro', 
                        'exerciseangia', 'slope', 'noofmajorvessels', 'target']
    
    print("\n" + "=" * 80)
    print("UNIVARIATE ANALYSIS - CATEGORICAL VARIABLES")
    print("=" * 80)
    
    for col in categorical_cols:
        if col in df.columns:
            print(f"\n{col.upper()}:")
            value_counts = df[col].value_counts()
            print(value_counts)
            print(f"Percentage: {value_counts / len(df) * 100}")

def visualize_distributions(df):
    """Visualize distributions"""
    numerical_cols = ['age', 'restingBP', 'serumcholestrol', 'maxheartrate', 'oldpeak']
    
    fig, axes = plt.subplots(len(numerical_cols), 2, figsize=(14, 4*len(numerical_cols)))
    
    for idx, col in enumerate(numerical_cols):
        if col in df.columns:
            # Histogram
            axes[idx, 0].hist(df[col].dropna(), bins=30, color='steelblue', alpha=0.7)
            axes[idx, 0].set_title(f'Histogram: {col}')
            axes[idx, 0].set_xlabel(col)
            axes[idx, 0].set_ylabel('Frequency')
            
            # Box plot
            axes[idx, 1].boxplot(df[col].dropna())
            axes[idx, 1].set_title(f'Box Plot: {col}')
            axes[idx, 1].set_ylabel(col)
    
    plt.tight_layout()
    plt.savefig('../../results/univariate_analysis.png')
    plt.close()

def main():
    """Main function"""
    df = load_data()
    univariate_numerical(df)
    univariate_categorical(df)
    visualize_distributions(df)
    print("\nUnivariate analysis complete!")

if __name__ == "__main__":
    main()

